library/server.c

```diff
@@ -183,10 +178,8 @@ struct dnet_node *dnet_server_node_create(struct dnet_config_data *cfg_data)
 	pthread_sigmask(SIG_SETMASK, &previous_sigset, NULL);
 	return n;
 
-	dnet_set_need_exit(n);
+err_out_cleanup_backends:
 	dnet_backends_cleanup_all(n);
-err_out_state_destroy:
-	dnet_state_put(n->st);
```

Comment:
Why was it removed?

Reply:
Because it is double free. If `dnet_state_create` successfully created `dnet_net_state`, net thread or `dnet_io_cleanup_states` will do last `dnet_state_put`.

Reply:
If state is destroyed here, it will be removed from `n->storage_state_list` and won't be processed by `dnet_io_cleanup_states()`. I found the only place where net thread calls `dnet_state_put()` - https://github.com/interiorem/elliptics/blob/master/library/pool.c#L963 and it is an error case. It should be carefully checked that there is no memory leak with this change.